File: gongwen-drafting/scripts/gongwen-agent/vision_ocr.m
#import <Foundation/Foundation.h>
#import <Vision/Vision.h>
#import <ImageIO/ImageIO.h>
#import <CoreGraphics/CoreGraphics.h>

static NSString *RecognizeText(NSString *imagePath) {
    NSURL *url = [NSURL fileURLWithPath:imagePath];
    CGImageSourceRef source = CGImageSourceCreateWithURL((__bridge CFURLRef)url, NULL);
    if (!source) {
        return @"";
    }
    CGImageRef image = CGImageSourceCreateImageAtIndex(source, 0, NULL);
    CFRelease(source);
    if (!image) {
        return @"";
    }

    VNRecognizeTextRequest *request = [[VNRecognizeTextRequest alloc] initWithCompletionHandler:nil];
    request.recognitionLevel = VNRequestTextRecognitionLevelAccurate;
    request.usesLanguageCorrection = YES;
    if (@available(macOS 11.0, *)) {
        request.recognitionLanguages = @[@"zh-Hans", @"zh-Hant", @"en-US"];
    }

    VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithCGImage:image options:@{}];
    NSError *error = nil;
    BOOL ok = [handler performRequests:@[request] error:&error];
    CGImageRelease(image);
    if (!ok) {
        if (error) {
            fprintf(stderr, "Vision OCR failed for %s: %s\n",
                    [imagePath UTF8String], [[error localizedDescription] UTF8String]);
        }
        return @"";
    }

    NSMutableArray<NSString *> *lines = [NSMutableArray array];
    for (VNRecognizedTextObservation *observation in request.results) {
        NSArray<VNRecognizedText *> *candidates = [observation topCandidates:1];
        if ([candidates count] > 0) {
            NSString *line = candidates[0].string;
            if ([line length] > 0) {
                [lines addObject:line];
            }
        }
    }
    return [lines componentsJoinedByString:@"\n"];
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        for (int i = 1; i < argc; i++) {
            NSString *path = [NSString stringWithUTF8String:argv[i]];
            printf("===PAGE:%s===\n", [[path lastPathComponent] UTF8String]);
            NSString *text = RecognizeText(path);
            printf("%s\n", [text UTF8String]);
        }
    }
    return 0;
}

File: gongwen-drafting/scripts/gongwen-agent/vision_ocr.swift
import Foundation
import Vision
import ImageIO
import CoreGraphics

func recognizeText(in imagePath: String) -> String {
    let url = URL(fileURLWithPath: imagePath)
    guard let source = CGImageSourceCreateWithURL(url as CFURL, nil),
          let image = CGImageSourceCreateImageAtIndex(source, 0, nil) else {
        return ""
    }

    let request = VNRecognizeTextRequest()
    request.recognitionLevel = .accurate
    request.usesLanguageCorrection = true
    if #available(macOS 11.0, *) {
        request.recognitionLanguages = ["zh-Hans", "zh-Hant", "en-US"]
    }

    let handler = VNImageRequestHandler(cgImage: image, options: [:])
    do {
        try handler.perform([request])
    } catch {
        fputs("Vision OCR failed for \(imagePath): \(error)\n", stderr)
        return ""
    }

    return (request.results ?? [])
        .compactMap { $0.topCandidates(1).first?.string }
        .joined(separator: "\n")
}

let paths = Array(CommandLine.arguments.dropFirst())
for path in paths {
    print("===PAGE:\((path as NSString).lastPathComponent)===")
    print(recognizeText(in: path))
}
